Grant API auth before MC connect

mikkamp opened the pull request:
### Changes proposed in this Pull Request:

This PR removes the requirement to connect the Merchant Center account first, before being able to grant the WPCOM API authentication. The following steps were implemented in this PR:

- Remove Merchant Center ID when requesting redirect URL
- Adjust Connection Test to support granting auth permissions
- Adjust unit tests to support the same


### Detailed test instructions:

#### Test API Pull grant for already existing users:
1. Start with a site that is fully onboarded (before using this PR)
2. Switch to this PR so we can test granting permissions
3. Go to the Google for WooCommerce > Settings page
4. Click on the "Get early access" button to start the flow to grant permissions
5. Confirm the process completes

> [!Note]
> There isn't really any way to confirm that the code is passing the Merchant ID here. So the best we can do is temporarily add some troubleshooting code to confirm it's included:
```diff
			$client   = $this->container->get( Client::class );
+			error_log( 'SDI Auth endpoint: ' . $this->get_sdi_auth_endpoint() );
			$result   = $client->get( $this->get_sdi_auth_endpoint() );
```
Make sure to have error logging enabled to a file so we can check the log entry, ex:
```
[20-Feb-2025 16:38:29 UTC] SDI Auth endpoint: https://api.woocommerce.com/google/google-sdi/v1/credentials/partners/WOO_COMMERCE/merchants/domain.test/oauth/redirect:generate?merchant_id=112113145
```

#### Test API Pull grant for new users:
1. Start with a site where everything is disconnected
2. Make sure to leave the WP.com connection connected, or reconnect it
3. We don't have the UI ready so we will test granting the API Pull auth through the connection test page
4. Go to the section "Partner API Pull Integration" and click the Grant access button
5. Go through the flow (if you have the logging enabled from above it should record the same URL without a `merchant_id`
6. Complete the flow and make sure we don't get any errors
7. Go through the onboarding and connect the MC account in the process
8. Go back to the connection test page and check the completed steps for the MC account, it should include the step `sdi_update`
![image](https://github.com/user-attachments/assets/0cea9f03-404d-4ef2-9ac6-7756c9dc6263)
9. Scroll down to the section "Partner API Pull Integration" and confirm everything is ready and approved
![image](https://github.com/user-attachments/assets/dd6da7e3-aad6-4e41-aafc-2c2ee4257f94)


### Additional details:
Project Thread: pcTzPl-2FJ-p2

### Changelog entry
* Tweak - Grant API auth before MC connect.

mikkamp committed:
Remove MC ID from auth endpoint request

mikkamp committed:
Allow filtering of return URL

mikkamp committed:
Add testing to grant API Pull at an earlier stage

mikkamp committed:
Adjust unit test to remove merchant_id

mikkamp committed:
Add request to update merchant account in SDI

mikkamp committed:
Add onboarding step to update merchant account in SDI

mikkamp committed:
Remove unused import

mikkamp committed:
Pass merchant ID with redirect generate if we've already connected

mikkamp committed:
Test redirect generate URL with and without MC ID

mikkamp committed:
Revise update function to use POST request

mikkamp committed:
Expect a mocked response to be sent once

mikkamp committed:
Tests for update_sdi_merchant_account

mikkamp committed:
Tests for account sdi_update

eason9487 commented on src/API/Google/Middleware.php:
```diff
@@ -598,4 +623,51 @@ public function get_sdi_auth_params() {
+					$result->getStatusCode()
+				);
+			}
+		} catch ( ClientExceptionInterface $e ) {
+			do_action( 'woocommerce_gla_guzzle_client_exception', $e, __METHOD__ );
+
+			throw new Exception(
+				$this->client_exception_message( $e, __( 'Error updating merchant account in Google Partner APP.', 'google-listings-and-ads' ) ),
+				$e->getCode()
+			);
```
After completing the WPCOM app authorization and connecting a GMC account, if I reconnect to the same/another GMC account, the connection fails with the following error.

![image](https://github.com/user-attachments/assets/efda9124-19ee-41b4-a2e0-3dc7ecd87730)

```json
{
  "message": "Error updating merchant account in Google Partner APP.",
  "id": 656614194
}
```

WC Error log

```
Error Automattic\WooCommerce\GoogleListingsAndAds\API\Google\Middleware::update_sdi_merchant_account Server error: `POST https://api.woocommerce.com/google/google-sdi/v1/credentials/partners/WOO_COMMERCE/merchants/eason-a8c.ngrok.dev/account:connect` resulted in a `500 Internal Server Error` response:
{
  "error": {
    "code": 500,
    "message": "Internal error encountered.",
    "status": "INTERNAL"
  }
}
```

I have to disconnect the WPCOM app (`DELETE /rest-api/authorize`) before I can reconnect to GMC.

mikkamp replied:
Discussed this with Google and they are looking into adding support for reconnecting a MC account. Currently no changes needed to our code.

eason9487 replied:
> Discussed this with Google and they are looking into adding support for reconnecting a MC account.

For a record, Google team has updated their API to support this, and we have confirmed the following scenarios work well now:
- Onboarding
   - Auth WPCom app without a connected GMC account
   - Connect to a GMC account
   - Reconnect to the same GMC account
   - Connect to another GMC account
- Settings
   - Auth WPCom app with a connected GMC account

eason9487 reviewed (commented):
Thanks for the work! The code looks good.

I tested it by locally merging the current `feature/onboarding-grant-api-pull` branch (0c355bacab8529ae580de13fc31a206e7f412be4) to this PR, and the happy paths worked well. However, I encountered errors when reconnecting Google Merchant Center account without disconnecting WPCOM app first. The error seems to come from Google API.

mikkamp commented:
Rebased and changed base branch to `feature/onboarding-grant-api-pull`

We are still waiting for Google to support reconnecting a Merchant Center account. But the code here is valid so we'll be merging it into the feature branch.
